src/main/scala/org/broadinstitute/hail/asm4s/FunctionBuilder.scala

```diff
@@ -88,110 +88,402 @@ abstract class FunctionBuilder[R](parameterTypeInfo: Array[TypeInfo[_]], returnT
+    val l = new java.util.concurrent.ConcurrentLinkedQueue[AbstractInsnNode]()
+    c.emit(l)
+    val dupes = l.groupBy(x => x).filter(_._2.toArray.length > 1).toSeq
+    if (!dupes.isEmpty)
+      println(s"dupes: $dupes")
+    l.foreach { x =>
+      mn.instructions.add(x)
+    }
     mn.instructions.add(new InsnNode(returnTypeInfo.returnOp))
     mn.instructions.add(end)
 
+    {
+      // print bytecode for debugging
+      val cwNoMaxesNoFrames = new ClassWriter(0)
+      cn.accept(cwNoMaxesNoFrames)
+      val cr = new ClassReader(cwNoMaxesNoFrames.toByteArray())
+      val tcv = new TraceClassVisitor(null, new Textifier, new PrintWriter(System.out))
+      cr.accept(tcv, 0)
+
+      // val sw = new StringWriter()
+      // CheckClassAdapter.verify(cr, false, new PrintWriter(sw))
+      // if (sw.toString().length() != 0) {
+      //   println("Verify Output for " + name + ":")
+      //   try {
+      //     val out = new BufferedWriter(new FileWriter("ByteCodeOutput.txt"))
+      //     out.write(sw.toString())
+      //     out.close()
+      //   } catch {
+      //     case e: IOException => System.out.println("Exception " + e)
+      //   }
+      //   println(sw)
+      //   println("Bytecode failed verification 1")
+      // }
+    }
+    {
+      val cw = new ClassWriter(ClassWriter.COMPUTE_MAXS + ClassWriter.COMPUTE_FRAMES)
+      cn.accept(cw)
+
+      val sw = new StringWriter()
+      CheckClassAdapter.verify(new ClassReader(cw.toByteArray()), false, new PrintWriter(sw))
+      if (sw.toString().length() != 0) {
+        println("Verify Output for " + name + ":")
+        try {
+          val out = new BufferedWriter(new FileWriter("ByteCodeOutput.txt"))
+          out.write(sw.toString())
+          out.close()
+        } catch {
+          case e: IOException => System.out.println("Exception " + e)
+        }
+        println(sw)
+        throw new IllegalStateException("Bytecode failed verification 2")
+      }
+    }
+
     val cw = new ClassWriter(ClassWriter.COMPUTE_MAXS +
       ClassWriter.COMPUTE_FRAMES)
     cn.accept(cw)
     val b = cw.toByteArray
-    val clazz = loadClass(null, b)
-
-    // print bytecode for debugging
-    /*
-    val cr = new ClassReader(b)
-    val tcv = new TraceClassVisitor(null, new Textifier, new PrintWriter(System.out))
-    cr.accept(tcv, 0)
-    */
-
-    clazz
-  }
-
-  def whileLoop(condition: Code[Boolean], body: Code[_]): Code[Unit] = {
-    val l1 = new LabelNode
-    val l2 = new LabelNode
-    new Code[Unit] {
-      def emit(il: InsnList): Unit = {
-        il.add(l1)
-        condition.emit(il)
-        il.add(new LdcInsnNode(0))
-        il.add(new JumpInsnNode(IF_ICMPEQ, l2))
-        body.emit(il)
-        il.add(new JumpInsnNode(GOTO, l1))
-        il.add(l2)
+    // val clazz = loadClass(null, b)
+
+    // clazz
+    b
+  }
+}
+
+class Function0Builder[R >: Null](implicit rti: TypeInfo[R]) extends FunctionBuilder[R](Array[TypeInfo[_]](), rti) {
```

Comment:
NB: `X >: Null` permits `X` which are reference types. `X <: AnyRef` includes `Nothing`, the empty type.

Unfortunately, certain traits are neither super types of `Null` nor subtypes of `AnyRef`, for example, `SeqLike`.